articles/types/jsdoc.md

```diff
@@ -1,3 +1,179 @@
+Moreover, if the result of `createAnimal` is assigned to a variable, the return type of createAnimal propogates to the assigned value.
+
+Finally, the text after the parameter names are *descriptions*, which show up as part of *quickinfo* anytime you hover over a construct that is documented
+
+<img src="../../images/jsdoc-quickinfo.png" width="1000px"/>
+
+and while filling in the paramters in a function call.
+
+<img src="../../images/jsdoc-param.png" width="1000px"/>
+
+Different constructs in JavaScript can use different forms of documentation, as described below.
+
+Visual Studio offers a rich editing experience inside JsDoc comments. The editor offers auto-completion on tags, names, and types as well as auto-indentation. You can use F12 and Shift+F12 for Go-to-definition and Find-All-References, respectively, as well hovering for quickinfo.
+
+## Supported Tags
+
+Any tag may be used for documentation, but Visual Studio treats some tags specially to improve the editor's ability to catch errors and help with refactoring.
+The following tags are handled specially. Each tag is followed by a description and example.
+
+* `@param` -- describes a parameter to a callable construct (eg: a function). Type annotations and descriptions are optional.
+
+```ts
+/**
+ * Repeats a string a certain number of times.
+ * Also: the future is bright and consists of native string repetition:
+ * https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/repeat
+ *
+ * @param {string} string  String to repeat
+ * @param {number} times   Number of times to repeat string. Integer.
+ * @see http://jsperf.com/string-repeater/2
+ */
+function repeatString(string, times) {
+  if (times === 1) {
+    return string;
+  }
+  if (times < 0) { throw new Error(); }
+  var repeated = '';
+  while (times) {
+    if (times & 1) {
+      repeated += string;
+    }
+    if (times >>= 1) {
+      string += string;
+    }
+  }
+  return repeated;
+}
+```
+
+* `@return` -- describes the return value of a callable construct (eg: a function). Type annotations and descriptions are optional.
+
+```ts
+/**
+ * Determine if the string ends with the specified substring.
+ *
+ * @param {string} haystack String to search in
+ * @param {string} needle   String to search for
+ * @return {boolean}
+ */
+function endsWith(haystack, needle) {
+  return haystack.slice(-needle.length) === needle;
+}
+```
+
+* `@class` -- tells Visual Studio to treat the next construct as a constructor. (Only available in JavaScript files)
+
+```ts
+var Cell = /** @class */ (function () {
+    function Cell() {
+    }
+    return Cell;
+}());
+```
+
+* `@augments` and `@extends` -- in JsDoc for class-like declarations, specifies which class the upcoming construct extends. Note that `@augments` and `@extends` are synonyms.
+
+```ts
+// @Filename: declarations.d.ts
+ declare class Thing<T> {
```

Comment:
This is a good example, except that it relies on two concepts not introduced in this document:

1. d.ts in a js project
2. generics

And it doesn't point out that the reason for the redundant jsdoc is actually that you can't specify type arguments in javascript.

I'm not sure whether the correct fix is to add the explanation, move it after some explanation, or to swap in a simpler example.

Reply:
prefixed with a warning.

I considered writing the following more-basic example, but it doesn't seem to work:

"
Like the `@class` tag above, these tags are useful with pre-ES6 javacsript, as in the following example:

```ts
function Base() {
    this.x = 10;
}

/** @augments Base */
function Derived() {
    Base.call(this);
    this.y = "hello";
}

let d = new Derived;
d.x;
```
"